src/object/srv_ec_aggregate.c

```diff
@@ -560,31 +806,779 @@ agg_update_vos(struct ec_agg_entry *entry, bool write_parity)
+	struct ec_agg_extent	*extent;
+	int			*status;
+	uint8_t			*bit_map = NULL;
+	uint8_t			 fcbit_map[OBJ_TGT_BITMAP_LEN] = {0};
+	uint8_t			 tbit_map[OBJ_TGT_BITMAP_LEN] = {0};
+	unsigned int		 len = entry->ae_oca->u.ec.e_len;
+	unsigned int		 k = entry->ae_oca->u.ec.e_k;
+	unsigned long            ss;
+	unsigned int		 i, full_cell_cnt = 0;
+	unsigned int		 cell_cnt = 0;
+	unsigned int		 estart, elen = 0;
+	unsigned int		 eend = 0;
+	int			 rc = 0;
+
+	/* For each contiguous extent, constructable from the extent list,
+	 * determine how many full cells, and how many cells overall,
+	 * are contained in the constructed extent.
+	 */
+	ss = k * len * entry->ae_cur_stripe.as_stripenum;
+	estart = entry->ae_cur_stripe.as_offset;
+	d_list_for_each_entry(extent, &entry->ae_cur_stripe.as_dextents,
 			      ae_link) {
-		daos_epoch_range_t	 epoch_range = { 0 };
-		int			 erc = 0;
-
-		epoch_range.epr_lo = epoch_range.epr_hi = agg_extent->ae_epoch;
-
-		erc = vos_obj_array_remove(agg_param->ap_cont_handle,
-					   entry->ae_oid, &epoch_range,
-					   &entry->ae_dkey, &entry->ae_akey,
-					   &agg_extent->ae_orig_recx);
-		if (erc)
-			D_ERROR("vos_obj_array_remove failed: "DF_RC"\n",
-				DP_RC(erc));
-		if (!rc && erc)
-			rc = erc;
-		ext_counter++;
-		if (ext_counter >= ext_total)
-			break;
+		D_ASSERT(!extent->ae_hole);
+		if (extent->ae_epoch <= entry->ae_par_extent.ape_epoch)
+			continue;
+		if (estart == extent->ae_recx.rx_idx - ss) {
+			eend = estart + extent->ae_recx.rx_nr;
+			elen = extent->ae_recx.rx_nr;
+			continue;
+		}
+		if (extent->ae_recx.rx_idx - ss > eend) {
+			cell_cnt += agg_count_cells(fcbit_map, tbit_map, estart,
+						    elen, k, len, entry->
+						    ae_cur_stripe.as_stripenum,
+						    &full_cell_cnt);
+			estart = extent->ae_recx.rx_idx;
+			elen = 0;
+		}
+		elen += extent->ae_recx.rx_nr;
+		eend += extent->ae_recx.rx_nr;
+	}
+	cell_cnt += agg_count_cells(fcbit_map, tbit_map, estart, elen, k, len,
+				    entry->ae_cur_stripe.as_stripenum,
+				    &full_cell_cnt);
+
+	if (full_cell_cnt >= k / 2 || cell_cnt == k) {
+		stripe_ud.asu_recalc = true;
+		cell_cnt = full_cell_cnt;
+		bit_map = fcbit_map;
+	} else
+		bit_map = tbit_map;
+
+	D_ASSERT(cell_cnt);
+	rc = agg_prep_sgl(entry);
+	if (rc)
+		goto out;
+	if (cell_cnt)
+		/* cell_cnt is zero if all cells are partial filled */
+		rc = agg_fetch_local_extents(entry, bit_map, cell_cnt,
+					     stripe_ud.asu_recalc);
+	if (rc)
+		goto out;
+
+	if (stripe_ud.asu_recalc) {
+		for (i = 0; i < k; i++) {
+			if (isset(bit_map, i))
+				clrbit(bit_map, i);
+			else
+				setbit(bit_map, i);
+		}
+	}
+
+	stripe_ud.asu_agg_entry = entry;
+	stripe_ud.asu_bit_map = bit_map;
+	stripe_ud.asu_cell_cnt = cell_cnt;
+
+	rc = ABT_eventual_create(sizeof(*status), &stripe_ud.asu_eventual);
+	if (rc != ABT_SUCCESS) {
+		rc = dss_abterr2der(rc);
+		goto out;
+	}
+	rc = dss_ult_create(agg_process_partial_stripe_ult, &stripe_ud,
+			    DSS_ULT_EC, 0, 0, NULL);
+	if (rc)
```

Comment:
If partial_stripe_ult will do remote fetch and calculating parity, then probably it should be put in a different xstream as IO.
Otherwise it does not need create a new ULT.

Reply:
I only create ULTs when there is a wait on data communication or usage of the ISA-L. The partial stripe ULT does both.

Reply:
I mean if it is only for waiting on data communication, then you do not have to create a ULT, unless you might consider using helper xstream.  Hmm, maybe you should, it looks like partial_stripe_ult() only do remote fetch and parity calculation.  
Anyway, this can be refined in the following patches.

Reply:
It is the helper XSTREAM. (DSS_ULT).